### DISCUSSION TITLE
Missing execute permission on „jre/bin/java“ file in Linux JRE

### DISCUSSION THREAD

**Discussion by willym74** [eclipse-justj/justj#30] [Q&A] [answered]
Hi, I’m using JustJ JRE21 in my RCP app, which is built with tycho for Windows and Linux environment. On Windows it works fine. On Linux I get on the app start: “JVM terminated. Exit code=13”, probably due to the missing “x” permission on the file “_plugins/org.eclipse.justj.openjdk.hotspot.jre.full.linux.x86_64_21.0.4.v20240802-1551/jre/bin/java_”. If I modify permissions manually (_chmod u+x_), the app starts properly.

In the “_p2.inf_” file of the corresponding JRE plugin I found the following:
```
instructions.install = \
  org.eclipse.equinox.p2.touchpoint.eclipse.setJvm(jvm:${artifact.location}/jre/bin/java);\
  org.eclipse.equinox.p2.touchpoint.eclipse.chmod(targetDir:${artifact.location},targetFile:jre/bin,permissions:+x,options:-R);\
  ...
```
So, actually all files under “_jre/bin/_” should have the execute permission, right? What goes wrong? 

![2024-09-24 15_31_25-Clipboard](https://github.com/user-attachments/assets/738ec773-7298-461f-91ce-29037dd9ff2a)

My “.target” file contains following:
```
<location includeAllPlatforms="false" includeConfigurePhase="true" includeMode="planner" includeSource="true" type="InstallableUnit">
    <unit id="org.eclipse.justj.openjdk.hotspot.jre.full.feature.group" version="0.0.0"/>
    <repository location="https://download.eclipse.org/justj/jres/21/updates/release/21.0.4"/>
</location>
```
The root "_pom.xml_":
```
<build>
  </pluginManagement>
    <plugins>
      …
      <plugin>
        <groupId>org.eclipse.tycho</groupId>
        <artifactId>target-platform-configuration</artifactId>
        <configuration>
          <executionEnvironment>org.eclipse.justj.openjdk.hotspot.jre.full-${jdk.version}</executionEnvironment>
        </configuration>
      </plugin>
      …
    </plugins>
  <pluginManagement>
  
  <plugins>
    …
    <plugin>
      <groupId>org.eclipse.tycho</groupId>
      <artifactId>target-platform-configuration</artifactId>
      <version>${tycho-version}</version>
      <configuration>
        <target>
          <artifact>
            <groupId>com.harman.spot.sps</groupId>
            <artifactId>com.harman.spot.sps.target</artifactId>
            <version>${revision}</version>
            <classifier>${target-platform}</classifier>
          </artifact>
        </target>
        <resolveWithExecutionEnvironmentConstraints>true</resolveWithExecutionEnvironmentConstraints>
        <environments>
          <environment>
            <os>win32</os>
            <ws>win32</ws>
            <arch>x86_64</arch>
          </environment>
          <environment>
            <os>linux</os>
            <ws>gtk</ws>
            <arch>x86_64</arch>
          </environment>
        </environments>            
        <dependency-resolution>
          <profileProperties>
            <org.eclipse.justj.buildtime>true</org.eclipse.justj.buildtime>
          </profileProperties>
        </dependency-resolution>
      </configuration>
    </plugin>
    …
    </plugins>
  </pluginManagement>
</build>
```

**Comment by merks** [comment]
I wonder how you are packaging up and distributing the product?

These things also build in include JustJ, so you can compere the permissions you see on these:

- https://www.eclipse.org/downloads/download.php?file=/oomph/products/latest/eclipse-inst-jre-linux64.tar.gz
- https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/2024-09/R/eclipse-committers-2024-09-R-linux-gtk-x86_64.tar.gz

I've never heard of someone having the problem you describe which makes me suspicions about how you package up the result perhaps losing the linux permissions...

I also assume you are building on Linux, not on Windows (because of course on Windows you cannot set these bits).

**Comment by willym74** [comment]
Thank you for your reply Ed! I tested apps you linked: both works fine and all permissions are set properly. 

Meanwhile I think, this is not a JustJ issue, but a Tycho issue. There is something what I'm missing in my build. Btw. the launcher file for the Linux distribution doesn't have the execute permission as well. I run the build on my Windows laptop just with
`mvn clean install`
Then I'm testing the Linux product distribution on the Ubuntu 20.04 in my local VM.

For the packaging of the product I'm just using the following tycho plugins in the product's pom.xml:
```
    <build>
        <plugins>
            <plugin>
                <groupId>org.eclipse.tycho</groupId>
                <artifactId>tycho-p2-director-plugin</artifactId>
                <version>${tycho-version}</version>
                
                <executions>
                    <execution>
                        <id>materialize-products</id>
                        <goals>
                            <goal>materialize-products</goal>
                        </goals>
                    </execution>
                    <execution>
                        <id>archive-products</id>
                        <goals>
                            <goal>archive-products</goal>
                        </goals>
                        <configuration>
                            <formats>
                                <linux>tar.gz</linux>
                                <win32>zip</win32>
                            </formats>
                        </configuration>
                    </execution>
                </executions>
                <configuration>
                    <products>
                        <product>
                            <id>${productId}</id>
                            <rootFolder>${rootFolder}</rootFolder>
                            <archiveFileName>${archiveFileName}</archiveFileName>
                        </product>
                    </products>
                    <source>targetPlatform</source>
                </configuration>
            </plugin>
        </plugins>
    </build>
```
I think, I have to investigate, what is missing in my build for setting chmod on all executable files.

**Reply by merks** [reply, marked as answer]
The touchpoint to set the execute flag is a no-op on Windows so you cannot build a Linux or macOS product on Windows.  So save the investigation time and instead get the same build working on Linux. Good luck. I think you have everything correctly in place.

**Comment by willym74** [comment]
Ok, that sounds plausible. My expectation was different because I previously (Oxygen) built directly from the product editor for both platforms: Windows and Linux and it worked great. Recently we updated to the latest Eclipse platform (2024_06) where building from the product editor is no longer supported and I had to switch it to Maven/Tycho. But, anyway, I will try to run the build process on a Linux machine, that should solve the problem.

Thanks for your support, I really appreciate it.
